Use radix tree for wildcard targets and map for plain targets

myleshorton opened the pull request:
This also changes serialization to use a simple array. Significantly simpler overall I think, and it loads much, much faster on startup!

myleshorton committed:
lots of changes to use a radix tree for wildcard targets and to just serialize all rulesets in an array

myleshorton committed:
cleanup

myleshorton commented:
FYI @oxtoacart whenever you get to it.

oxtoacart commented on deserializer.go:
```diff
@@ -4,9 +4,12 @@ import (
 	"time"
 
 	"github.com/getlantern/golog"
+	iradix "github.com/hashicorp/go-immutable-radix"
```
Out of curiosity, why did you choose an immutable radix tree instead of the mutable version?

myleshorton replied:
Basically just because by default I think everything should be immutable to reduce overall state, and in this case post-startup there are never any changes to the tree.

oxtoacart replied:
I was just asking because immutable implementations often have a bit of a performance penalty. IIRC the main use of "immutable" data structures is actually for state that you want to be able to mutate but allow old references to the state to continue to be able to use the old version without locking. In this particular case, we're not actually interested in that sort of benefit since, as you say, we don't change anything post-startup. Would be interesting to do a quick microbenchmark to see if there's a significant memory or cpu penalty vs https://github.com/armon/go-radix.

oxtoacart commented on httpse.go:
```diff
@@ -0,0 +1,143 @@
+		return "", false
+	}
+
+	start := time.Now()
```
I would use mtime here.

oxtoacart commented on httpse.go:
```diff
@@ -0,0 +1,143 @@
+	start := time.Now()
+	if val, ok := h.plainTargets.Load().(map[string]*ruleset)[url.Host]; ok {
+		r, hit := h.rewriteWithRuleset(url, val)
+		h.addTiming(time.Now().Sub(start), url.String())
```
You could put this in a defer to avoid repeating it.

oxtoacart commented on httpse.go:
```diff
@@ -0,0 +1,143 @@
+		h.addTiming(time.Now().Sub(start), url.String())
+		return r, hit
+	}
+	_, val, match := h.wildcardTargets.Load().(*iradix.Tree).Root().LongestPrefix([]byte(url.Host))
```
Don't we need to check `rewriteWithRuleset` for both the suffix and prefix match?

myleshorton replied:
Yeah both fall through to call `rewriteWithRuleset` below.

oxtoacart replied:
As written, if there's a matching suffix rule we will only check `rewriteWithRuleset` for that suffix rule. If that returns false, we won't rewrite, even though there may be a matching prefix rule that would result in a rewrite.

To check both, you'd need to fall through on line 69 if `hit` is false.

myleshorton replied:
Oh I see what you're saying. Will fix!

oxtoacart commented on httpse.go:
```diff
@@ -0,0 +1,143 @@
+	}
+	h.statM.Unlock()
+
+	h.statM.RLock()
```
You can avoid this second set of locking operations by simply saving to local variables in the above block and printing out from those.

myleshorton replied:
Yeah that was a little messy. I ended up switching to a channel on a separate go routine.

oxtoacart commented:
@myleshorton Finished first round of review.

myleshorton committed:
added processing of other rulesets in cases where there is no match, switched to mtime, and switched to channels for stats

myleshorton commented:
OK @oxtoacart I think I addressed all of your comments except the benchmark. The rules load in about 250ms on my machine, so I don't think performance is a huge issue at this stage, but I'm curious myself so will go ahead and run 'em!

myleshorton committed:
switched to mutable radix tree since the API is slighlty cleaner

myleshorton commented:
Hey I switched over to the mutable version. The benchmarks were pretty much identical, but the API for the mutable version's a little cleaner, so I just stuck with it @oxtoacart.

oxtoacart commented on httpse.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func (h *httpse) addTiming(t *timing) {
+	ms := t.dur.Nanoseconds() / int64(time.Millisecond)
```
I actually liked the old way better. Amongst other things, this has the potential for leaking goroutines since this thing doesn't have a Close method.

oxtoacart replied:
I think something like the below would be cleanest and also more efficient than using a channel:

```golang
func (h *httpse) addTiming(dur time.Duration, host string) {
 	ms := dur.Nanoseconds() / int64(time.Millisecond)
 	h.statM.Lock()
 	h.runs++
 	h.totalTime += ms
 	if ms > h.max {
 		h.max = ms
 		h.maxHost = host
    }
    runs, totalTime, max, maxHost := h.runs, h.totalTime, h.max, h.maxHost
 	h.statM.Unlock()
  
 	log.Debugf("Average running time: %vms", float64(totalTime/runs))
 	log.Debugf("Max running time: %vms for host: %v", max, maxHost)
 }
```

oxtoacart replied:
BTW, if you use the above approach, you don't need a RWMutex, just a regular Mutex.

myleshorton replied:
What makes you say it's more efficient @oxtoacart? More efficient for the recording of the stat, or more efficient for the caller? I'm definitely more concerned with the latter, as to me under load the lock contention here could get reasonably significant. This is also one of those relatively rare go routines that actually does run for the lifetime of the application, so one of the rare ones where we don't actually want to close it.

oxtoacart replied:
Channels are not lock-free, they use locks too, see https://github.com/golang/go/blob/master/src/runtime/chan.go#L49. By using a channel, you end up having to create this `timing` struct, which not only clutters the code but also incurs a little creation and garbage collection overhead. The efficiency difference is not great, but using a channel plus goroutine certainly doesn't improve performance and IMHO hurts readability in this case.

oxtoacart replied:
Mostly just for my own edification, I created some benchmarks with high concurrency to compare the two approaches - https://github.com/getlantern/httpseverywhere/blob/simplerox/timing_bench_test.go

Results from three runs show that the locking approach is consistently faster, from anywhere between 18 to 33 %.

In the scheme of things, 50 microseconds per op really isn't that much overhead so it's not a big deal either way, and the difference is small enough that we could effectively consider the performance "equivalent". This is however interesting for me in the general case, as it suggests that the decision between channels vs locks should generally be driven by code structure/readability/robustness rather than performance.

```
➜  httpseverywhere git:(simplerox) ✗ go test -v -run blah -bench . | grep -v "^DEBUG"
BenchmarkNoMatch-4   	DEBUG httpse: httpse.go:155 Average running time: 0ms
   30000	     52212 ns/op
BenchmarkChannel-4   	DEBUG channel: timing_bench_test.go:45 Average running time: 3000ms
   30000	     60727 ns/op
BenchmarkLock-4      	DEBUG channel: timing_bench_test.go:32 Average running time: 21000ms
   30000	     40311 ns/op
PASS
ok  	github.com/getlantern/httpseverywhere	9.038s
➜  httpseverywhere git:(simplerox) ✗ go test -v -run blah -bench . | grep -v "^DEBUG"
BenchmarkNoMatch-4   	DEBUG httpse: httpse.go:155 Average running time: 0ms
   30000	     55284 ns/op
BenchmarkChannel-4   	DEBUG channel: timing_bench_test.go:45 Average running time: 11000ms
   30000	     50108 ns/op
BenchmarkLock-4      	DEBUG channel: timing_bench_test.go:32 Average running time: 4000ms
   30000	     40269 ns/op
PASS
ok  	github.com/getlantern/httpseverywhere	8.667s
➜  httpseverywhere git:(simplerox) go test -v -run blah -bench . | grep -v "^DEBUG"
BenchmarkNoMatch-4   	DEBUG httpse: httpse.go:155 Average running time: 0ms
   30000	     51739 ns/op
BenchmarkChannel-4   	DEBUG channel: timing_bench_test.go:45 Average running time: 3000ms
   30000	     47314 ns/op
BenchmarkLock-4      	DEBUG channel: timing_bench_test.go:32 Average running time: 1000ms
   30000	     38461 ns/op
PASS
ok  	github.com/getlantern/httpseverywhere	8.123s

oxtoacart replied:
BTW, I agree that the leaking goroutine isn't a big deal given how we'll use this library, I just noticed that the API effectively allows for the creation of an arbitrary number of HTTPSEverywhere instances and doesn't provide the ability to close them even though they do leak resources.

myleshorton replied:
Hey so my point on the lock contention is that the relevant code here is not the entire printing of the timings but rather the impact on the performance of HTTPSEverywhere itself -- i.e. on the impacts of adding HTTPEverywhere to any library. So the advantage of using a channel is that you offload everything to another go routine -- the locking, the logging, etc, while minimizing the impact to whatever go routine is using HTTPSEverywhere in the first place.

Here's a simplified benchmark that shows the channels significantly beating locking from this perspective -- you actually can't even get the channel approach to register at all:


```
$ go test -v -run blah -bench . | grep -v "^DEBUG"
(standard input):2:BenchmarkNoMatch-8   	DEBUG httpse: httpse.go:155 Average running time: 0ms
(standard input):120205:   50000	     28656 ns/op
(standard input):120208:BenchmarkChannel-8   	2000000000	         0.00 ns/op
(standard input):122209:BenchmarkLock-8      	DEBUG channel: timing_bench_test.go:31 Average running time: 0ms
(standard input):134209:2000000000	         0.03 ns/op
```

As you bump up the concurrency the effect is more pronounced of course -- at 10000:

```
$ go test -v -run blah -bench . | grep -v "^DEBUG"
(standard input):2:BenchmarkNoMatch-8   	DEBUG httpse: httpse.go:155 Average running time: 0ms
(standard input):120206:   50000	     31027 ns/op
(standard input):120208:BenchmarkChannel-8   	2000000000	         0.00 ns/op
(standard input):140209:BenchmarkLock-8      	DEBUG channel: timing_bench_myles_test.go:31 Average running time: 0ms
(standard input):600209:1000000000	         0.67 ns/op
```

myleshorton replied:
I ended up just putting that benchmark here to avoid conflicts:

https://github.com/getlantern/httpseverywhere/blob/simpler/timing_bench_test.go

oxtoacart replied:
Interesting example. I think what you're saying is that as long as the throughput of the processing goroutine exceeds the rate at which data is submitted to the channel, submitters will block for a negligible amount of time. I think that's a good point. In my test, the rate of submission exceeded the rate of processing, in which case sends to a channel don't behave much differently than just locking.

In the case of HTTPSEverywhere, generally the rate of processing should far exceed the rate of submission. If you wanted to make it robust to temporary high activity, you could use a buffered channel to absorb the temporary spikes. Probably not worth it though.

I'm cool with merging this if you are.

myleshorton replied:
Yeah that's right, although my point is also partly that the performance of some go routines is more important than the performance of others, so offloading via a channel seems like a useful approach in those cases.

oxtoacart commented on httpse.go:
```diff
@@ -0,0 +1,157 @@
+		return "", false
+	}
+
+	start := mtime.Now()
```
You can also use `mtime.Stopwatch()` if you want, which was designed for this exact purpose.

oxtoacart commented:
Other than the stats thing, this LGTM. I don't see the stats thing as a blocker, but I liked it better the way that it was before (sans extra locking call).

oxtoacart commented on deserializer.go:
```diff
@@ -19,50 +22,104 @@ func newDeserializer() *deserializer {
+	plains := make(map[string]*ruleset)
+	wildcards := radix.New()
+	for _, rs := range rulesets {
+		wildcards = d.addRuleset(rs, plains, wildcards)
```
Now that you're not using the immutable version, you don't need to keep returning the wildcards tree.

myleshorton replied:
Fixed!

oxtoacart commented:
Done with another round of review. Other than the potential for leaking goroutines, we're firmly into nitpicking territory now.

myleshorton committed:
more code review tweaks

myleshorton committed:
updated benchmark

oxtoacart committed:
Fixed problem with locking/channel benchmark

oxtoacart commented on timing_bench_test.go:
```diff
@@ -46,7 +47,9 @@ func BenchmarkChannel(b *testing.B) {
 	for i := 0; i < concurrency; i++ {
 		go func() {
-			ch <- &timing{host: "myhost", dur: dur}
+			for j := 0; j < int(math.Ceil(float64(b.N)/concurrency)); j++ {
```
@myleshorton The reason why you couldn't get the benchmark to register a non-zero number is that you ignored the `b.N` parameter. That parameter determines the number of "ops" that the benchmark is expecting to run, and it calculates the time per op by dividing the runtime by `b.N`.

After adding back in code to make sure we actually run approximately `b.N` iterations, we get these results:

```
➜  httpseverywhere git:(simpler) ✗ go test -v -run blah -bench . | grep -v "^DEBUG"
BenchmarkNoMatch-4   	DEBUG httpse: httpse.go:155 Average running time: 0ms
   30000	     49316 ns/op
BenchmarkChannel-4   	 2000000	       626 ns/op
BenchmarkLock-4      	DEBUG channel: timing_bench_test.go:32 Average running time: 0ms
   30000	     47065 ns/op
PASS
ok  	github.com/getlantern/httpseverywhere	12.987s
```

As expected, simply submitting new stat structs to the channel is pretty cheap.

oxtoacart replied:
If I make the test channel a buffered channel with depth 1000, the time to submit goes up slightly but is still tiny. This tells me that a buffered channel is probably worth doing here.

```
➜  httpseverywhere git:(simpler) ✗ go test -v -run blah -bench . | grep -v "^DEBUG"
BenchmarkNoMatch-4   	DEBUG httpse: httpse.go:156 Max running time: 0ms for host:
   30000	     51011 ns/op
BenchmarkChannel-4   	 2000000	       774 ns/op
BenchmarkLock-4      	DEBUG channel: timing_bench_test.go:32 Average running time: 0ms
   50000	     36920 ns/op
PASS
ok  	github.com/getlantern/httpseverywhere	12.619s
```

myleshorton replied:
Oh great improvement @oxtoacart on the benchmark thanks. I went ahead and made the channel buffered at 100, which I agree seems better. That whole exercise was super useful actually, so glad we went down that path. I feel like for really large concurrent applications that ability to offload less critical things to other go routines is likely pretty vital.

myleshorton committed:
made channel buffered
